File: services/attendance-service/src/test/java/com/visor/school/attendance/integration/AttendanceReportIntegrationTest.java
package com.visor.school.attendance.integration;

import com.visor.school.attendance.model.AttendanceStatus;
import com.visor.school.attendance.repository.AttendanceRepository;
import com.visor.school.attendance.service.AttendanceService;
import com.visor.school.attendance.service.AttendanceService.AttendanceReport;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.amqp.rabbit.core.RabbitTemplate;

import java.time.LocalDate;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Integration test for attendance report generation
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class AttendanceReportIntegrationTest {

    @Autowired
    private AttendanceService attendanceService;

    @Autowired
    private AttendanceRepository attendanceRepository;

    @MockBean
    private RabbitTemplate rabbitTemplate;

    private final UUID testClassId = UUID.randomUUID();
    private final UUID testTeacherId = UUID.randomUUID();

    @BeforeEach
    void setup() {
        // Create attendance records for testing
        LocalDate startDate = LocalDate.now().minusDays(7);

        // Create records for multiple days
        for (int dayOffset = 0; dayOffset <= 7; dayOffset++) {
            LocalDate date = startDate.plusDays(dayOffset);
            attendanceService.markAttendanceDirectly(
                UUID.randomUUID(),
                testClassId,
                date,
                AttendanceStatus.PRESENT,
                testTeacherId,
                null
            );
            attendanceService.markAttendanceDirectly(
                UUID.randomUUID(),
                testClassId,
                date,
                AttendanceStatus.ABSENT,
                testTeacherId,
                null
            );
        }
    }

    @Test
    void shouldGenerateAttendanceReportForClass() {
        // Given
        LocalDate startDate = LocalDate.now().minusDays(7);
        LocalDate endDate = LocalDate.now();

        // When
        AttendanceReport report = attendanceService.generateClassReport(testClassId, startDate, endDate);

        // Then
        assertNotNull(report);
        assertEquals(testClassId, report.getClassId());
        assertEquals(startDate, report.getStartDate());
        assertEquals(endDate, report.getEndDate());
        assertNotNull(report.getAttendanceRate());
        assertTrue(report.getAttendanceRate().getRate() >= 0.0);
        assertTrue(report.getAttendanceRate().getRate() <= 100.0);
    }

    @Test
    void shouldCalculateCorrectAttendanceStatistics() {
        // Given
        LocalDate startDate = LocalDate.now().minusDays(7);
        LocalDate endDate = LocalDate.now();

        // When
        AttendanceReport report = attendanceService.generateClassReport(testClassId, startDate, endDate);

        // Then
        // Note: The Kotlin test accessed fields like report.totalRecords, report.presentCount.
        // These likely map to properties on AttendanceReport or AttendanceRate.
        // In my Java structure: report.getAttendanceRate().getTotal(), .getPresent(), etc.
        
        assertTrue(report.getAttendanceRate().getTotal() > 0);
        assertTrue(report.getAttendanceRate().getPresent() >= 0);
        assertTrue(report.getAttendanceRate().getAbsent() >= 0);
        assertTrue(report.getAttendanceRate().getLate() >= 0);
        assertTrue(report.getAttendanceRate().getExcused() >= 0);
    }

    @Test
    void shouldHandleEmptyDateRange() {
        // Given
        LocalDate futureStartDate = LocalDate.now().plusDays(30);
        LocalDate futureEndDate = LocalDate.now().plusDays(37);

        // When
        AttendanceReport report = attendanceService.generateClassReport(testClassId, futureStartDate, futureEndDate);

        // Then
        assertNotNull(report);
        assertEquals(0, report.getAttendanceRate().getTotal());
        assertEquals(0.0, report.getAttendanceRate().getRate());
    }
}

File: services/attendance-service/src/test/kotlin/com/visor/school/attendanceservice/integration/AttendanceReportIntegrationTest.kt
package com.visor.school.attendanceservice.integration

import com.visor.school.attendanceservice.model.AttendanceStatus
import com.visor.school.attendanceservice.repository.AttendanceRepository
import com.visor.school.attendanceservice.service.AttendanceService
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.util.UUID

/**
 * Integration test for attendance report generation
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class AttendanceReportIntegrationTest @Autowired constructor(
    private val attendanceService: AttendanceService,
    private val attendanceRepository: AttendanceRepository
) {

    private val testClassId = UUID.randomUUID()
    private val testTeacherId = UUID.randomUUID()

    @BeforeEach
    fun setup() {
        // Create attendance records for testing
        val startDate = LocalDate.now().minusDays(7)
        val endDate = LocalDate.now()

        // Create records for multiple days
        for (dayOffset in 0..7) {
            val date = startDate.plusDays(dayOffset.toLong())
            attendanceService.markAttendanceDirectly(
                studentId = UUID.randomUUID(),
                classId = testClassId,
                date = date,
                status = AttendanceStatus.PRESENT,
                markedBy = testTeacherId
            )
            attendanceService.markAttendanceDirectly(
                studentId = UUID.randomUUID(),
                classId = testClassId,
                date = date,
                status = AttendanceStatus.ABSENT,
                markedBy = testTeacherId
            )
        }
    }

    @Test
    fun `should generate attendance report for class`() {
        // Given
        val startDate = LocalDate.now().minusDays(7)
        val endDate = LocalDate.now()

        // When
        val report = attendanceService.generateClassReport(testClassId, startDate, endDate)

        // Then
        assertNotNull(report)
        assertEquals(testClassId, report.classId)
        assertEquals(startDate, report.startDate)
        assertEquals(endDate, report.endDate)
        assertNotNull(report.attendanceRate)
        assertTrue(report.attendanceRate >= 0.0)
        assertTrue(report.attendanceRate <= 1.0)
    }

    @Test
    fun `should calculate correct attendance statistics`() {
        // Given
        val startDate = LocalDate.now().minusDays(7)
        val endDate = LocalDate.now()

        // When
        val report = attendanceService.generateClassReport(testClassId, startDate, endDate)

        // Then
        assertNotNull(report.totalRecords)
        assertTrue(report.totalRecords > 0)
        assertNotNull(report.presentCount)
        assertNotNull(report.absentCount)
        assertNotNull(report.lateCount)
        assertNotNull(report.excusedCount)
    }

    @Test
    fun `should handle empty date range`() {
        // Given
        val futureStartDate = LocalDate.now().plusDays(30)
        val futureEndDate = LocalDate.now().plusDays(37)

        // When
        val report = attendanceService.generateClassReport(testClassId, futureStartDate, futureEndDate)

        // Then
        assertNotNull(report)
        assertEquals(0, report.totalRecords)
        assertEquals(0.0, report.attendanceRate)
    }
}

